src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.GeneratorDriverInitializationCache.cs

```diff
@@ -77,31 +82,18 @@ GeneratorDriver CreateGeneratorDriverAndRunGenerators(CancellationToken cancella
 
                 return generatorDriver.RunGenerators(compilation, generatorFilter, cancellationToken);
             }
-        }
-
-        public void EmptyCacheForProjectsThatHaveGeneratorDriversInSolution(SolutionCompilationState state)
-        {
-            // If we don't have any cached drivers, then just return before we loop through all the projects
-            // in the solution. This is to ensure that once we hit a steady-state case of a Workspace's CurrentSolution
-            // having generators for all projects, we won't need to keep anything further in our cache since the cache
-            // will never be used -- any running of generators in the future will use the GeneratorDrivers already held by
-            // the Solutions.
-            //
-            // This doesn't need to be synchronized against other mutations to _driverCache. If we see it as empty when
-            // in reality something was just being added, we'll just do the cleanup the next time this method is called.
-            if (_driverCache.IsEmpty)
-                return;
 
-            foreach (var (projectId, tracker) in state._projectIdToTrackerMap)
+            GeneratorDriver UpdateDriverAndRunGenerators(GeneratorDriver driver)
             {
-                if (tracker.GeneratorDriver is not null)
-                    EmptyCacheForProject(projectId);
+                driver = UpdateGeneratorDriverToMatchState(driver, projectState);
+
+                return driver.RunGenerators(compilation, generatorFilter, cancellationToken);
             }
         }
 
-        public void EmptyCacheForProject(ProjectId projectId)
+        public void UpdateCacheWithForGeneratorDriver(GeneratorDriver driver)
         {
-            ImmutableInterlocked.TryRemove(ref _driverCache, projectId, out _);
+            _driverCache = AsyncLazy.Create(driver);
```

Comment:
Cute. Definitely doc this field to explain this will happen.  Ah. I see you did doc this well